### PR TITLE
[Feature](merge-on-write)Support ignore mode for merge-on-write unique table

### PR DESCRIPTION
**Pull request by bobhan1**
## Proposed changes

This PR add support for ignore mode for merge-on-write unique table. It can be used in native insert statement(`insert ignore into ...`) or stream load. When the ignore mode is enabled, if the inserted row with key which exists in the table,  the row data will be dicarded. If a inserted row with key which doesn't exist in table, the row data will be inserted. 
The ignore mode is only available in mow unique table without sequence column and can't be used in a partial update.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 54.47 seconds
 stream load tsv:          503 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          28.8 seconds inserted 10000000 Rows, about 347K ops/s
 storage size: 17170840524 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230713011400_clickbench_pr_177520.html

**Comment by bobhan1** [comment]
run p0

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 53.23 seconds
 stream load tsv:          502 seconds loaded 74807831229 Bytes, about 142 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          28.9 seconds inserted 10000000 Rows, about 346K ops/s
 storage size: 17170075550 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230713103658_clickbench_pr_177635.html

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 51 seconds
 stream load tsv:          502 seconds loaded 74807831229 Bytes, about 142 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17169589207 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230713113153_clickbench_pr_177724.html

**Comment by bobhan1** [comment]
run feet

**Comment by bobhan1** [comment]
run feut

File: docs/zh-CN/docs/sql-manual/sql-reference/Data-Manipulation-Statements/Manipulation/INSERT.md [hunk 1]
```diff
@@ -44,7 +44,8 @@ INSERT INTO table_name
 ```
 
  Parameters
-
+> IGNORE: insert ignore模式，仅当目标表为开启merge-on-write的unique表时有效。开启后，对于插入的行，如果该行的key在表中不存在，则插入该行数据。如果key在表中不存在，则丢弃这行数据。
```
Comment (morningman) [inline docs/zh-CN/docs/sql-manual/sql-reference/Data-Manipulation-Statements/Manipulation/INSERT.md:47]: 这个ignore 语义看上出不是很明确，建议用一个明确的语法

Comment (zhannngchen) [reply docs/zh-CN/docs/sql-manual/sql-reference/Data-Manipulation-Statements/Manipulation/INSERT.md:47]: https://dev.mysql.com/doc/refman/8.0/en/insert.html
@morningman 这个是参考mysql的语法的

File: be/src/olap/rowset/segment_v2/segment_writer.cpp [hunk 2]
```diff
@@ -587,6 +598,17 @@ Status SegmentWriter::append_block(const vectorized::Block* block, size_t row_po
                         << "found duplicate key or key is not sorted! current key: " << key
                         << ", last key" << last_key;
                 RETURN_IF_ERROR(_primary_key_index_builder->add_item(key));
+                if (is_unique_key_ignore_mode) {
+                    RowLocation loc;
+                    RowsetSharedPtr rowset;
+                    auto status = _tablet->lookup_row_key(key, false, specified_rowsets, &loc,
```
Comment (zhannngchen) [inline be/src/olap/rowset/segment_v2/segment_writer.cpp:604]: Don't check if key exists here, you should change the behavior of all the places of update delete bitmap

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 51.48 seconds
 stream load tsv:          550 seconds loaded 74807831229 Bytes, about 129 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          28.9 seconds inserted 10000000 Rows, about 346K ops/s
 storage size: 17163498312 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230714004533_clickbench_pr_178193.html

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 55.28 seconds
 stream load tsv:          508 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17160911020 Bytes

**Comment by bobhan1** [comment]
run feut

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 50.74 seconds
 stream load tsv:          505 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17156091863 Bytes

**Review by zhannngchen — COMMENTED** [review]
1. should consider delete sign
2. should consider partial update, partial update calculate delete bitmap in `append_block_with_partial_content`

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.23 seconds
 stream load tsv:          508 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17170899357 Bytes

**Comment by bobhan1** [comment]
run p0

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 43.67 seconds
 stream load tsv:          506 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17171144325 Bytes

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.15 seconds
 stream load tsv:          512 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17162051883 Bytes

**Comment by bobhan1** [comment]
run arm

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 48.25 seconds
 stream load tsv:          515 seconds loaded 74807831229 Bytes, about 138 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17162327376 Bytes

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run clickbench

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.68 seconds
 stream load tsv:          513 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17162167442 Bytes

**Comment by lide-reed** [comment]
Can you add a parameter and let stream load support ignore?  @bobhan1

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.71 seconds
 stream load tsv:          538 seconds loaded 74807831229 Bytes, about 132 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17161938948 Bytes

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.26 seconds
 stream load tsv:          543 seconds loaded 74807831229 Bytes, about 131 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          66 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17162162876 Bytes

**Comment by bobhan1** [comment]
run p0

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.89 seconds
 stream load tsv:          539 seconds loaded 74807831229 Bytes, about 132 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17162184717 Bytes

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.44 seconds
 stream load tsv:          543 seconds loaded 74807831229 Bytes, about 131 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          66 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17162241043 Bytes

**Comment by yuxuan-luo** [comment]
在 stream load 使用 format = json 时,插入会直接删除掉数据

**Comment by bobhan1** [comment]
> 在 stream load 使用 format = json 时,插入会直接删除掉数据

you can describe your problem more explicitly in `Discussions`

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run clickbench

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.43 seconds
 stream load tsv:          534 seconds loaded 74807831229 Bytes, about 133 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          68 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          28.9 seconds inserted 10000000 Rows, about 346K ops/s
 storage size: 17162000775 Bytes

**Review by yuxuan-luo — APPROVED** [review]
LGTM

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.24 seconds
 stream load tsv:          530 seconds loaded 74807831229 Bytes, about 134 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17162102118 Bytes

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

File: be/src/olap/tablet.cpp [hunk 3]
```diff
@@ -2951,42 +2953,50 @@ Status Tablet::calc_segment_delete_bitmap(RowsetSharedPtr rowset,
...
-                // delete bitmap will be calculate when memtable flush and
-                // publish. The two stages may see different versions.
-                // When there is sequence column, the currently imported data
-                // of rowset may be marked for deletion at memtablet flush or
-                // publish because the seq column is smaller than the previous
-                // rowset.
-                // just set 0 as a unified temporary version number, and update to
-                // the real version number later.
+            if (UNLIKELY(is_unique_key_ignore_mode)) {
+                if (st.is<OK>() || st.is<KEY_ALREADY_EXISTS>()) {
```
Comment (zhannngchen) [inline be/src/olap/tablet.cpp:2957]: lookup_row_key will return KEY_ALREADY_EXISTS only when the table has seq column. You should ignore seq column if `is_unique_key_ignore_mode` is true (set the second parameter of `lookup_row_key` as false)

File: be/src/olap/rowset/beta_rowset_writer.cpp [hunk 4]
```diff
@@ -131,7 +131,8 @@ Status BetaRowsetWriter::add_block(const vectorized::Block* block) {
     return _segment_creator.add_block(block);
 }
 
-Status BetaRowsetWriter::_generate_delete_bitmap(int32_t segment_id) {
+Status BetaRowsetWriter::_generate_delete_bitmap(int32_t segment_id,
+                                                 bool is_unique_key_ignore_mode) {
```
Comment (zhannngchen) [inline be/src/olap/rowset/beta_rowset_writer.cpp:135]: it's not necessary to add such additional parameter, you can use `_context.tablet_schema->is_unique_key_ignore_mode()` directly at L153?

Comment (zhannngchen) [inline be/src/olap/rowset/beta_rowset_writer.cpp:135]: it's not necessary to add these parameters, you can get the tablet schema inside `calc_segment_delete_bitmap`.

File: be/src/olap/rowset/beta_rowset_writer.cpp [hunk 5]
```diff
@@ -714,7 +715,8 @@ Status BetaRowsetWriter::add_segment(uint32_t segment_id, SegmentStatistics& seg
         }
     }
     if (_context.mow_context != nullptr) {
-        RETURN_IF_ERROR(_generate_delete_bitmap(segment_id));
+        RETURN_IF_ERROR(_generate_delete_bitmap(
+                segment_id, _context.tablet_schema->is_unique_key_ignore_mode()));
```
Comment (zhannngchen) [inline be/src/olap/rowset/beta_rowset_writer.cpp:719]: You also need to consider publish phase.

File: docs/zh-CN/docs/sql-manual/sql-reference/Data-Manipulation-Statements/Manipulation/INSERT.md [hunk 6]
```diff
@@ -44,7 +44,8 @@ INSERT INTO table_name
 ```
 
  Parameters
-
+> IGNORE: insert ignore模式，仅当目标表为开启merge-on-write的unique表时有效。开启后，对于插入的行，如果该行的key在表中不存在，则插入该行数据。如果key在表中不存在，则丢弃这行数据。当目标表中存在sequence列时无法通过insert ignore语句进行插入操作。
```
[the remark this answers is not in the document]
Comment (bobhan1) [reply docs/zh-CN/docs/sql-manual/sql-reference/Data-Manipulation-Statements/Manipulation/INSERT.md:47]: 我觉得这这个也直接在fe禁掉？

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

File: fe/fe-core/src/test/java/org/apache/doris/planner/OlapTableSinkTest.java [hunk 7]
```diff
@@ -206,7 +206,7 @@ public void testListPartition(
         };
 
         OlapTableSink sink = new OlapTableSink(dstTable, tuple, Lists.newArrayList(p1.getId()), false);
-        sink.init(new TUniqueId(1, 2), 3, 4, 1000, 1, false, false);
+        sink.init(new TUniqueId(1, 2), 3, 4, 1000, 1, false, false, false);
```
Comment (zhannngchen) [inline fe/fe-core/src/test/java/org/apache/doris/planner/OlapTableSinkTest.java:209]: Is it possiable to add sume UT to cover the logics you added in OlatTableSink?

Comment (bobhan1) [reply fe/fe-core/src/test/java/org/apache/doris/planner/OlapTableSinkTest.java:209]: the regression tests have covered them

**Review by zhannngchen — DISMISSED** [review]
LGTM

**Comment by bobhan1** [comment]
run p0

**Review by liaoxin01 — APPROVED** [review]
LGTM

**Comment by bobhan1** [comment]
run buildall

**Comment by zhannngchen** [comment]
run buildall

**Review by zhannngchen — APPROVED** [review]
LGTM

**Pull request merged**